186274265 Synchronous Case Highlighting for Tables, Datadecks, and Graphs

tealefristoe opened the pull request:
PT Stories:
https://www.pivotaltracker.com/story/show/186274228
https://www.pivotaltracker.com/story/show/186274265
https://www.pivotaltracker.com/story/show/186274374

This PR adds highlighting for cases in table and datacard tiles (both single card and sort views).
- It's now possible to select a case in a datacard tile.
- Highlighting a case switches that case's card to be displayed in single card view in datacard tiles.
- Clicking anywhere on a card in sort view highlights the case (this is according to the spec, which seems wrong to me, which is why I'm pointing it out).
- Highlighting in graphs works, but the styling isn't to spec. Dennis will be handling this in another PR.

One major under the hood change in this PR is to use `dataSet.selection` for table case selection instead of `stores.sharedSelection`.

Another change is adding default graph settings to `app-config.json`. Someone with more knowledge of the graph should take a look at these settings to make sure they're correct.

Finally, there are a bunch of formatting issues and code optimizations in this PR, as well as a fix for a linting error that was introduced in another PR.

Note for QA: It's currently not possible to clear a table's column selection when selecting a case in another tile, which is requested in the PT story. This will be handled in a future story, when we address attribute selection.

dennisrcao commented:
@tealefristoe
I've gone ahead and cleaned up this PR for any `console.logs`
Due to the growing complexity of the highlighting ticket - we've decided that this PR should not include that ticket, but instead address Data Card syncing.

tealefristoe commented:
@dennisrcao I ended up going back and removing all changes to the graph code from this PR. A few formatting changes were happening automatically because of our linting rules, but otherwise the PR doesn't touch the graph code. I think this will make it easier for reviewers to understand what are relevant changes vs incomplete changes that will be finished in a later PR.

tealefristoe commented:
@kswenson Thanks for your patience with this PR. I think it's finally ready for a look over when you have some time. One cypress test is still failing but it looks like a usual failure unrelated to this work.

kswenson commented on src/components/tiles/table/use-grid-context.ts:
```diff
@@ -49,15 +51,18 @@ export const useGridContext = ({ modelId, showRowLabels, triggerColumnChange }:
+    if (select !== dataSet.isCaseSelected(rowId)) {
+      clearSelection({ row: false });
+      dataSet.selectCases([rowId], !dataSet.isCaseSelected(rowId));
+      triggerColumnChange();
```
```suggestion
      triggerRowChange();
```
This probably isn't entirely clear from the naming, but `triggerRowChange()` is meant to trigger rendering of rows, while `triggerColumnChange()` is meant to trigger a re-evaluation of the column definitions (which should then result in re-rendering the entire table). In the CODAP v3 table this is now handled in a way that makes it easier to only re-render the affected rows rather than having to re-render all rows.

kswenson commented on src/components/tiles/table/use-grid-context.ts:
```diff
@@ -49,15 +51,18 @@ export const useGridContext = ({ modelId, showRowLabels, triggerColumnChange }:
-  }, [clearColumnSelection, modelId, sharedSelection]);
+    clearSelection();
+    dataSet.setSelectedCases([rowId]);
+    triggerColumnChange();
```
```suggestion
    triggerRowChange();
```
see comment above

kswenson commented on src/components/tiles/table/use-grid-context.ts:
```diff
@@ -75,9 +80,11 @@ export const useGridContext = ({ modelId, showRowLabels, triggerColumnChange }:
-    sharedSelection.setSelected(modelId, Array.from(_rows) as string[]);
-  }, [modelId, sharedSelection]);
+    dataSet.setSelectedCases(Array.from(_rows) as string[]);
+    triggerColumnChange();
```
Like the others above, if this is necessary it should be `triggerRowChange()`, but it may not be necessary at all. This is handling the case of a selection which started with the table so RDG already knows exactly which rows have changed and so I would think that it wouldn't be necessary to trigger an external re-render of the table.

kswenson reviewed (changes requested):
👍 Looking good -- I made a number of suggestions but most of them are relatively local/minor. Seeing some of this old CLUE table code I'm wishing there were time for a larger synchronization of code from the CODAP table which has a number of refinements, but I guess those will have to wait for another day.

kswenson commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -40,6 +40,14 @@ export const DataCardToolComponent: React.FC<ITileProps> = observer(function Dat
+  // When the highlighted case is set, show it
+  const selectedCaseId = content.dataSet.firstSelectedCaseId;
+  useEffect(() => {
+    if (selectedCaseId) {
+      content.setCaseIndex(content.dataSet.caseIndexFromID(selectedCaseId));
+    }
```
This potential undo/redo issue could also be deferred in the interest of expediency.

tealefristoe replied:
I tried playing around with this a bit but I wasn't sure what the undo/redo problem is. `setCaseIndex` calls `withoutUndo`, so it doesn't get added to the stack. And `dataset.selection` is volatile, so changing it isn't undoable.

tealefristoe commented:
@kswenson Thanks for the feedback, there were a lot of good simplifications in there 👍 

I addressed all your suggestions except the undo/redo one. I think this might actually not be a problem, but let me know if I misunderstood you and I'll take another look.

kswenson reviewed (approved):
👍Looks good -- I missed the withoutUndo() on the selection.

tealefristoe commented:
@tejal-shah These are good observations, I'll try to address them today and will let you know how they go.

tealefristoe commented:
@tejal-shah I believe I was able to address both of the issues you uncovered. Let me know if anything doesn't work as expected 👍

tejal-shah commented:
Looks great!
